// @evershop/product_review/dist/pages/frontStore/productView/Reviews.js
import PropTypes from "prop-types";
import React from "react";
import "./Reviews.scss";
import { _ } from "@evershop/evershop/lib/locale/translate/_";
import Rating from "../../../components/Rating";
export default function Reviews({ product: { reviews = [] } }) {
    return (React.createElement("div", { id: "product__reviews", className: "mt-2" },
        React.createElement("h3", null, _("Customer Reviews")),
        React.createElement("ul", { className: "review__list" },
            reviews.length === 0 && (React.createElement("li", { className: "flex flex-col gap-1" }, _("Be the first to review this product"))),
            reviews.map((review) => (React.createElement("li", { key: review.uuid, className: "flex flex-col gap-1" },
                React.createElement("div", { className: "rating" },
                    React.createElement(Rating, { rating: review.rating })),
                React.createElement("p", { className: "comment" }, review.comment),
                React.createElement("div", null,
                    React.createElement("span", null, _("Review by ${customer_name}", {
                        customer_name: review.customerName,
                    })))))))));
}
Reviews.propTypes = {
    product: PropTypes.shape({
        reviews: PropTypes.arrayOf(PropTypes.shape({
            rating: PropTypes.number.isRequired,
            comment: PropTypes.string.isRequired,
            customerName: PropTypes.string.isRequired,
            createdAt: PropTypes.string.isRequired,
        })),
    }).isRequired,
};
export const layout = {
    areaId: "productPageMiddleLeft",
    sortOrder: 45,
};
export const query = `
  query {
    product(id: getContextValue("productId")) {
      reviews {
        reviewId
        uuid
        rating
        customerName
        comment
        createdAt
      }
    }
  }
`;
